[Bug] Sampled expectation does not support density matrices

### Short description

When feeding a density matrix and calling `expectation`, we obtain an error. This is due to not handling the densitymatrix case in the api. 

### What is the expected result?

_No response_

### What is the actual result?

_No response_

### Steps/Code to reproduce

Any program such that:

```
# test density mat
dm = density_mat(state) 
exp_ad_dm = expectation(circ, dm, values, obs, DiffMode.AD) # works
exp_ad_sampled_dm = expectation(
    circ,
    dm,
    values,
    obs,
    DiffMode.AD,
    n_shots=100000,
) # different error can appear
```

### Tracebacks (optional)

If added in the `test_sampled_diff` test, we obtain:

```
FAILED tests/test_differentiation.py::test_sampled_diff[dtype0-Z-2-3] - RuntimeError: einsum(): subscript c is repeated for operand 0 but the sizes don't match, 1 != 2
FAILED tests/test_differentiation.py::test_sampled_diff[dtype1-Y-2-3] - RuntimeError: einsum(): subscript c is repeated for operand 0 but the sizes don't match, 1 != 2
FAILED tests/test_differentiation.py::test_sampled_diff[dtype0-Y-2-5] - IndexError: index 3 is out of bounds for axis 0 with size 3
FAILED tests/test_differentiation.py::test_sampled_diff[dtype1-Z-1-5] - IndexError: index 3 is out of bounds for axis 0 with size 3
FAILED tests/test_differentiation.py::test_sampled_diff[dtype1-Z-1-3] - RuntimeError: einsum(): subscript c is repeated for operand 0 but the sizes don't match, 1 != 2
FAILED tests/test_differentiation.py::test_sampled_diff[dtype0-Z-1-3] - RuntimeError: einsum(): subscript c is repeated for operand 0 but the sizes don't match, 1 != 2
FAILED tests/test_differentiation.py::test_sampled_diff[dtype0-Y-1-3] - RuntimeError: einsum(): subscript c is repeated for operand 0 but the sizes don't match, 1 != 2
FAILED tests/test_differentiation.py::test_sampled_diff[dtype1-Y-1-3] - RuntimeError: einsum(): subscript c is repeated for operand 0 but the sizes don't match, 1 != 2
FAILED tests/test_differentiation.py::test_sampled_diff[dtype0-Z-2-5] - IndexError: index 3 is out of bounds for axis 0 with size 3
FAILED tests/test_differentiation.py::test_sampled_diff[dtype1-Y-1-5] - IndexError: index 3 is out of bounds for axis 0 with size 3
FAILED tests/test_differentiation.py::test_sampled_diff[dtype1-Y-2-5] - IndexError: index 3 is out of bounds for axis 0 with size 3
FAILED tests/test_differentiation.py::test_sampled_diff[dtype1-Z-2-5] - IndexError: index 3 is out of bounds for axis 0 with size 3
FAILED tests/test_differentiation.py::test_sampled_diff[dtype0-Z-1-5] - IndexError: index 3 is out of bounds for axis 0 with size 3
FAILED tests/test_differentiation.py::test_sampled_diff[dtype0-Y-2-3] - RuntimeError: einsum(): subscript c is repeated for operand 0 but the sizes don't match, 1 != 2
FAILED tests/test_differentiation.py::test_sampled_diff[dtype0-Y-1-5] - IndexError: index 3 is out of bounds for axis 0 with size 3
FAILED tests/test_differentiation.py::test_sampled_diff[dtype1-Z-2-3] - RuntimeError: einsum(): subscript c is repeated for operand 0 but the sizes don't match, 1 != 2
```

### Environment details (optional)

_No response_

### Would you like to work on this issue?

Yes

--- a/pyqtorch/api.py
+++ b/pyqtorch/api.py
@@ -7,7 +7,7 @@ from logging import getLogger
 import torch
 from torch import Tensor
 
-from pyqtorch.apply import apply_operator
+from pyqtorch.apply import apply_operator, apply_operator_dm
 from pyqtorch.circuit import QuantumCircuit
 from pyqtorch.differentiation import (
     AdjointExpectation,
@@ -16,7 +16,7 @@ from pyqtorch.differentiation import (
 )
 from pyqtorch.embed import Embedding
 from pyqtorch.hamiltonians import Observable
-from pyqtorch.utils import DiffMode, sample_multinomial
+from pyqtorch.utils import DensityMatrix, DiffMode, sample_multinomial
 
 logger = getLogger(__name__)
 
@@ -166,13 +166,24 @@ def sampled_expectation(
         ).permute((2, 0, 1))
     )
     eigvals = eigvals.squeeze()
-    eigvec_state_prod = apply_operator(
-        state,
-        eigvecs.T.conj(),
-        tuple(range(n_qubits)),
-    )
-    eigvec_state_prod = torch.flatten(eigvec_state_prod, start_dim=0, end_dim=-2).t()
-    probs = torch.pow(torch.abs(eigvec_state_prod), 2)
+    if isinstance(state, DensityMatrix):
+        eigvec_state_prod = apply_operator_dm(
+            state,
+            eigvecs.T.conj(),
+            tuple(range(n_qubits)),
+        )
+        probs = torch.diagonal(eigvec_state_prod, dim1=0, dim2=1).real
+
+    else:
+        eigvec_state_prod = apply_operator(
+            state,
+            eigvecs.T.conj(),
+            tuple(range(n_qubits)),
+        )
+        eigvec_state_prod = torch.flatten(
+            eigvec_state_prod, start_dim=0, end_dim=-2
+        ).t()
+        probs = torch.pow(torch.abs(eigvec_state_prod), 2)
     if circuit.readout_noise is not None:
         batch_samples = circuit.readout_noise.apply(probs, n_shots)
 

--- a/tests/test_differentiation.py
+++ b/tests/test_differentiation.py
@@ -9,8 +9,10 @@ from pyqtorch import DiffMode, expectation
 from pyqtorch.matrices import COMPLEX_TO_REAL_DTYPES
 from pyqtorch.primitives import Primitive
 from pyqtorch.utils import (
+    ATOL,
     GRADCHECK_ATOL,
     GRADCHECK_ATOL_hamevo,
+    density_mat,
 )
 
 
@@ -156,6 +158,20 @@ def test_sampled_diff(
         n_shots=100000,
     )
     assert torch.allclose(exp_ad, exp_ad_sampled, atol=1e-01)
+
+    # test density mat
+    dm = density_mat(state)
+    exp_ad_dm = expectation(circ, dm, values, obs, DiffMode.AD)
+    exp_ad_sampled_dm = expectation(
+        circ,
+        dm,
+        values,
+        obs,
+        DiffMode.AD,
+        n_shots=100000,
+    )
+    assert torch.allclose(exp_ad_dm, exp_ad, atol=ATOL)
+    assert torch.allclose(exp_ad, exp_ad_sampled_dm, atol=1e-01)
 
 
 @pytest.mark.xfail  # Adjoint Scale is currently not supported

